diff --git a/nvbench/CMakeLists.txt b/nvbench/CMakeLists.txt
index eec4a15..d945ce6 100644
--- a/nvbench/CMakeLists.txt
+++ b/nvbench/CMakeLists.txt
@@ -38,7 +38,7 @@ set(srcs
 
 if (NVBench_ENABLE_CUPTI)
   list(APPEND srcs detail/measure_cupti.cu)
-  if (EXISTS "${nvbench_cupti_root}/include/cupti_profiler_host.h")
+  if (EXISTS "${NVBench_CUPTI_INCLUDE_DIR}/cupti_profiler_host.h")
     list(APPEND srcs cupti_profiler_host.cxx)
   else()
     list(APPEND srcs cupti_profiler_nvpw.cxx)
